[jello] Newsstand Week1

hjsong333 opened the pull request:
## 설명
1. 깔끔한 코드보다는 그리드 UI와 기능 구현에 집중했습니다.

    컴포넌트 생성자에서 내부에 필요한 모든 요소를 createElement로 생성하다보니 매우 비대해졌습니다. customCreateElement 따위의 유틸함수를 만들어서 최대한 정리하고 컴포넌트의 프로퍼티 중에서도 필요없는 것들은 정리하려고 합니다. 다른 아이디어 환영입니다.

2. main에서 모든 상태를 관리하고 props로 전달하는 과정에서 드릴링이 있습니다.

    이 부분은 불편하고 불필요하다고 생각이 들어서 추후 수정 계획이 있습니다. 아이디어가 있으시다면 코멘트 부탁드립니다.
    필요한 컴포넌트에서 필요한 상태만 받아쓸 수 있다면 좋겠는데 미뤄왔던 작업을 해야할 때가 온 것 같습니다! 🔥 

3. 정보나 도움이 필요하신 부분이 있다면 코멘트 부탁드립니다.

    가식적인 코멘트는 No라고 했지만, 정말로  괜찮아보여서 써보고 싶거나 궁금하다 싶은 부분이 있다면 코멘트 부탁드립니다.

## 구현 목록
- 그리드 보기 UI
- 1~4 페이지 넘기기
- 새로고침하면 언론사 섞기
- 각 셀에 마우스를 올리면, 구독하지 않은 경우에는 '구독하기', 구독한 경우에는 '해지하기'버튼이 보인다.

## 미리보기
### 페이지 넘기기 & 호버시 구독/해지 버튼

https://github.com/codesquad-members-2023/fe-max--newsstand/assets/60080167/79771f3a-39e7-4e60-9ed6-5123e82bd19a

### 새로고침 시 섞기

https://github.com/codesquad-members-2023/fe-max--newsstand/assets/60080167/20fc34e1-c302-4964-b7a3-7ce1e3d14666

## Dev log

https://skitter-tachometer-31f.notion.site/Week1-6dbada2593dc4b3a8edf0360e6517ec3

hjsong333 committed:
chore: 개발 환경 세팅

개발을 위한 도구와 환경 준비
- Vite
- TypeScript
- Prettier
- Sass

hjsong333 committed:
chore: set basic normalize, mixins, variables

- remove unused css and svg file.

hjsong333 committed:
style: 아이콘 파일 추가 및 헤더 레이아웃

hjsong333 committed:
docs: 프로젝트 개요 작성

hjsong333 committed:
feat: 뉴스스탠드 데이터 크롤링 코드 추가

youzysu commented on src/counter.ts:
```diff
@@ -0,0 +1,9 @@
+export function setupCounter(element: HTMLButtonElement) {
```
사용하시지 않는 파일은 바로 삭제해주시면 좋을 것 같아요!

hjsong333 committed:
chore: 크롤링 코드 디렉토리 이름 변경

hjsong333 committed:
docs: README 백로그 URL 수정

hjsong333 committed:
feat: change method of rendering header component

- With assignment of template literal, detailed re-render is hard.
- Result of experimental component rendering code is successful.

hjsong333 committed:
refactor: delete unused files and code

hjsong333 committed:
fix: Fix issue that hot reload not working

- `noEmit: true` option in tsconfig.json is reason.
- Delete `noEmit` and `allowImportingTsExtensions`.

hjsong333 committed:
refactor: noEmit, allowImportingTsExtensions option rollback

- Reason for hot reload issue is cache of vscode.

hjsong333 committed:
style: apply module css to components

hjsong333 committed:
style: replace scss with css and module css

- To reduce the cost of class naming(BEM).
- Instead of mixin and variable, use CSS variables and font/bg classes.

hjsong333 committed:
feat: display date injected from outside of Header

- To display date dynamically

hjsong333 committed:
feat: implement page refresh on logo click

- Change tag for logo from div to anchor.
- Add refresh handler for click event.

hjsong333 committed:
style: remove anchor's text-decoration and repaint

hjsong333 committed:
chore: move assets to public directory

- Because anyway assets will be copied to public.
- To set path for img or font can be easier.

hjsong333 committed:
feat: implement NewsBar component and render it

- No layout and just rendering

hjsong333 committed:
style: layout for NewsBar and RollingNews

hjsong333 committed:
refactor: move components and module.css files to components directory

hjsong333 committed:
feat: implement Main and MainHeader components

- Create tabs in MainHeader
- no layout just render elements

hjsong333 committed:
style: layout for MainHeader based on figma design

hjsong333 committed:
feat: create GridView component and render it

- no layout and features.
- just render with dummy data.

hjsong333 committed:
chore: edit page title

hjsong333 committed:
style: layout for GridView component

- Adjust NewsStand width to 930px(table width).
- Dummy logo src from naver news-stand.

hjsong333 committed:
feat: add inserting id to gridData in crawler

hjsong333 committed:
feat: display media logos on GridView

- Crawl image data from naver news-stand.
- Sending data through component structure(drilling)

hjsong333 committed:
feat: render grid images with current page state

hjsong333 committed:
style: create and render arrows in GridView

hjsong333 committed:
feat: implement page flipping of GridView

- Add click event handler to arrows.
- Event handler create action with different type.
- Each action change page state in main.ts.
- After change state, sending changed state to GridView as props.
- With props, change image source URL and alt of grid cells.

hjsong333 committed:
feat: implement media logo random shuffle

- Get new gridData when page is refreshed.
- Apply Fisher-Yates algorithm on getGridData function.

hjsong333 committed:
feat: subscribe button rendered when mouse over

- Generate `turnOnSubscriptionCover` action when mouse hover at each cell in grid.
- Change states about hover(isHover and hoverIndex).
- Send states to GridView and render related element.

hjsong333 committed:
feat: turn off subscription cover when mouse leave

- Mouse leave event is added to tbody in GridView.
- turnOffSubscriptionCover action has payload which contain only hoverOnGrid.
- After invoke, hoveredCellIndex turns to -1.

hjsong333 committed:
feat: switch subscription button text

- Based on subscribed media's id.
- To reuse subscription cover, create it when component constructed.

hjsong333 committed:
fix: not displayed plus icon of subscription

- Because of inappropriate use of `textContent`.
- `textContent` replace all children elements of subscription button.
- Replace textContent with textNode.

hjsong333 commented:
https://www.notion.so/Week1-6dbada2593dc4b3a8edf0360e6517ec3?pvs=4

altmit commented on src/components/main/GridView.ts:
```diff
@@ -0,0 +1,207 @@
+import style from './GridView.module.css';
+
+export default class GridView {
+  public element: HTMLElement;
+  private table: HTMLTableElement;
+  private tbody: HTMLTableSectionElement;
+  private rows: HTMLTableRowElement[];
+  private cells: HTMLTableCellElement[];
+  private subscription: HTMLDivElement;
+  private leftArrow: HTMLAnchorElement;
+  private rightArrow: HTMLAnchorElement;
+  private page: number;
```
constructor에서 element를 만들기 위해 이렇게 다 사전에 타입을 지정해야 하는 일이 생각보다 번거로울 것 같습니다.
만약 element가 더 많은 component라면 더 번거로울 것 같습니다.

이런 문제를 어떻게 해결할 수 있을까요?
같이 고민해 보고 의견을 공유하면 좋을 것 같습니다.

youzysu replied:
마침 저도 리뷰 달았던 부분인데, 이렇게 타입을 지정해주는 것과 아래 생성 시 타입 추론이 되도록만 해두는 것의 차이가 있을까요?

> 아래 생성 시에 타입스크립트가 알아서 타입 추론을 해주는 것으로 알고 있는데, 이렇게 명시적으로 다시 타입을 작성해주시는 이유가 있을까요? 매번 작성하기 번거로워서 저는 생략했던 부분이라 궁금해서 여쭤봅니다!

hjsong333 replied:
거창한 이유가 있어서 저렇게 한 것은 아니지만 제 생각을 말씀드리겠습니다.
다른 개발자 혹은 스스로 실수하는 것을 방지하기 위해서입니다.
타입이 동적으로 변할 수 있다면 어디선가 제가 예측하지 못하는 오류가 발생할 수 있는데 이걸 시스템/환경으로 못하게 제어하려고 사용하는거라 생각합니다.

```
class Foo {
  number: number;
  consctructor() {
    this.number = 1;
  }
  
  changeNumber() {
    this.number = '1'; // Type 'string' is not assignable to type 'number'.
  }
}
```

단적으로 이런 에러는 타입 어노테이션 없으면 안 알려주죠.

hjsong333 replied:
annotation 없으면 동적으로 타입이 변하는 줄 알아서 쓴거긴 한데, 안 써도 타입 추론 후에는 안 변하게 막는군요.
그래도 나중에 봤을 때 좀 덜 헷갈릴 것 같긴 하네요.

jsh3418 replied:
모두 프로퍼티로 가지고 있어야 하는 요소들인가요? 굳이 프로퍼티가 아니어도 되는 요소면 빼도 좋을 것 같다고 생각됩니다.

altmit commented on src/components/main/GridView.ts:
```diff
@@ -0,0 +1,207 @@
+    cellsInRow: number;
+  };
+
+  constructor({ gridInfo }: { gridInfo: GridInfo }) {
```
모든 element를 constructor에서 만들어서 constructor가 너무 길어진것 같습니다.
꼭 모든 element를 constructor에서 만들어야 하나요?

hjsong333 replied:
제가 생각해도 많이 길긴하네요. 
함수로 빼서 줄여볼 계획인데, 혹시 jay가 말씀하시는건 컴포넌트 생성 시점이 아닌 다른 타이밍에 element를 만들어주는 방향일까요?

altmit replied:
저도 단순 함수로 빼면 가독성이 좋아질 것 같다는 생각만 했습니다.

youzysu replied:
constructor 함수 내에서 줄바꿈으로 구분짓고 있긴 하지만 모든 하위 element를 생성하고, attribute 설정까지 이뤄지고 있는데 element 혹은 그보다는 큰 하위 컴포넌트 단위로 메서드로 나누면 어떨까요?

hjsong333 replied:
하위 컴포넌트를 만들어볼까도 고민해봐야겠네요. 
다만 드릴링이 더 깊어질 것 같아서 그게 좀 고민이긴 합니다.

altmit commented on src/components/main/GridView.ts:
```diff
@@ -0,0 +1,207 @@
+    for (let i = 0; i < this.cells.length; i++) {
+      const cell = this.cells[i];
+      if (isHover && hoverIndex === i) {
+        const mediaLogo = cell.firstElementChild?.firstElementChild! as HTMLElement;
+        const textNode = this.subscription.firstElementChild?.lastChild as HTMLElement;
```
as 보다 타입가드를 활용하면 더 좋을 것 같습니다!

hjsong333 replied:
아직 타입스크립트가 익숙치 않아서 되는대로 해둔건데 키워드 잘 받겠습니다 👍

altmit commented on src/components/main/GridView.ts:
```diff
@@ -0,0 +1,207 @@
+      this.leftArrow.classList.remove('no-display');
+    }
+
+    if (this.page === 3 && !this.rightArrow.classList.contains('no-display')) {
+      this.rightArrow.classList.add('no-display');
+    } else if (this.page !== 3 && this.rightArrow.classList.contains('no-display')) {
+      this.rightArrow.classList.remove('no-display');
+    }
```
3페이지가 넘는 경우가 생길 수 있지 않을까요?

hjsong333 replied:
아직 거기까지 고려하지는 못했네요. 리팩토링할 때 고려해보겠습니다.

altmit reviewed (commented):
젤로도 저와 비슷하게 남세의 flux 자료를 보고 흐름을 따라가고 있어 저와 방향성이 비슷한 것 같아 나중에 grid를 만드는 과정에 도움이 될 것 같습니다

silvertae commented on crawler/grid.js:
```diff
@@ -0,0 +1,41 @@
+      });
+    });
+
+    brandMarks.push(
+      ...data.map((d, index) => {
+        return {
+          id: brandMarks.length + index + 1,
+          ...d
+        };
+      })
+    );
```
제가 spread 연산자가 아직 익숙치 않아서 그런지 `...data.map()` 앞에 `...`이 붙어있어서 코드를 이해하는데 좀 어려웠네요. 저는 spread 한 data의 각 요소들에 map을 적용하는 건 줄 알고 순간 해석이 안됐네요.

`data.map()`이 다 실행되고 난 후 return 된 새로운 배열을 spread 해서 각 요소들을 brandMarks 배열에 push 해주는 코드였군요. 이런 식으로 쓸 수 있다는 건 새로 알게되었네요.

hjsong333 replied:
가독성이 안 좋아서 충분히 오해할 수 있는 부분이겠네요. 변수로 한 번 받아서 spread operator를 쓰는 방향으로 리팩토링 해보면 조금 나아질까 싶네요.

silvertae commented on src/utils/dataFetch.ts:
```diff
@@ -0,0 +1,9 @@
+import gridData from '../db/grid.json';
+
+export const getGridImgs = () => {
+  for (let i = gridData.length - 1; i > 0; i--) {
+    const j = Math.floor(Math.random() * (i + 1));
+    [gridData[i], gridData[j]] = [gridData[j], gridData[i]];
+  }
```
이 부분이 Fisher-Yates 알고리즘을 적용하신 부분이군요. 그런데 함수 이름만 봤을때는 이미지를 가져오는 역할을 한다기보다는 섞어주는 역할만 하고 있는 것 같아서, 함수나 파일이름을 약간 수정하면  이해가 더 잘 될 것 같아요. 뭔가 shuffle이나 random 이라는 단어가 들어가면 좋지 않을까요?

hjsong333 replied:
함수 이름과 안 어울리긴 하네요. 섞는 로직을 다른 함수로 분리하는걸 고려해봐야겠군요. 
지적 감사합니다! 👍

silvertae commented on src/components/header/Header.ts:
```diff
@@ -0,0 +1,50 @@
+
+  private getDate(date: Date): string {
+    const year = date.getFullYear();
+    const month = (date.getMonth() + 1).toString().padStart(2, '0');
+    const day = date.getDate().toString().padStart(2, '0');
```
`padStart()`라는 메서드는 처음 봤네요. 찾아보니 유용하게 사용할 수 있을 것 같아요. 새롭게 배워갑니다. 👍

silvertae reviewed (commented):
아직 컴포넌트 단위 개발과 state와 props의 차이점에 대해서 공부 중이어서 핵심적인 코드 부분에는 리뷰를 많이 못 달았네요. 
지금 constructor에서 creatElement로 일일이 요소들을 만들어서 렌더링 하고있어서 코드가 길어지는 부분을 어떻게 해결할지 저도 한 번 고민해보겠습니다!

youzysu commented on src/components/NewsStand.ts:
```diff
@@ -0,0 +1,37 @@
+    this.element.append(this.header.element, this.newsBar.element, this.main.element);
+  }
+
+  updateProps(props: NewNewsStandProps) {
```
이렇게 기존의 props와 비교하지 않고 바로 새로운 props로 updateProps를 해도 상태가 바뀐 부분만 적용되는지 궁금해요!

hjsong333 replied:
지금까지 구현한 기능에서는 굳이 비교하지 않아도 상태가 바뀔때만 updateProps가 호출되서 미처 고려하지 못했네요.
비교하는 로직을 넣어줘야겠습니다. 감사합니다. 👍

youzysu commented on src/components/main/GridView.ts:
```diff
@@ -0,0 +1,207 @@
+
+    this.tbody = document.createElement('tbody');
+
+    this.numbers = {
```
매직 넘버를 어떻게 없애야 하나 고민중이었는데 이렇게 묶어둘 수 있겠군요! 다만 numbers라는 네이밍은 조금 포괄적이라 직관적이진 않은 것 같기도 합니다 ㅎㅎ

hjsong333 replied:
맞습니다. 역할이 잘 이해 안 될 수 있을 것 같네요. 😅

youzysu commented on src/components/main/GridView.ts:
```diff
@@ -0,0 +1,207 @@
+  }
+
+  setEvent() {
+    this.cells.forEach((cell, index) => {
```
이벤트 위임으로 등록해주면 어떨까요? 추가로 어떤 이벤트 핸들러는 위의 constructor 함수 내에서 등록해주고 setEvent에서는 하나의 이벤트 핸들러만 등록하고 있어서 setEvent 메서드와 constructor 함수의 역할에 대해 좀 더 명확하게 정의해주시면 좋을 것 같아요.

hjsong333 replied:
이벤트 위임은 mouseover로 가능할 것 같은데, 불필요하게 이벤트가 너무 많이 발생해서 이런 방식으로 처리해줬습니다.
호버 이벤트 기능을 만들때 로직이 길어져서 빼낸건데, 다른 이벤트도 이쪽으로 빼줘야겠네요.

youzysu reviewed (commented):
이번 주 새롭게 배우신 패턴을 적용해서 재밌게(?) 구현하고 계시다는 젤로의 말씀이 넘 좋았습니다 😊
모든 HTML element를 생성하는 방식 가독성 측면에서 취약한만큼 의미 있는 함수 단위로 묶어주면 더 좋을 것 같아요. 이번 한주도 수고 많으셨습니다!

bakhacode commented on src/components/header/Header.ts:
```diff
@@ -0,0 +1,50 @@
+    this.element = document.createElement('header');
+    this.element.classList.add(style.header);
+
+    this.logo = document.createElement('a');
+    this.logo.classList.add(style.logo);
+    this.logo.href = '#';
+    this.logo.addEventListener('click', (event) => {
+      event.preventDefault();
+      location.reload();
+    });
```
엘리먼트별로 모든 동작을 모아놓은 게 일관성은 있지만 이벤트같은 경우는 뭔가 따로 빼서 두면 좋을 것 같아요! 물론 여기서는 어차피 하나지만 혹시 다음 진행에서 상태별로 이벤트 등록하는게 많아졌을 때는 중요부분이다보니 생성과 속성 설정 이후에 따로 모아주면 좀 더 읽을 때 빼먹지 않고 볼 수 있지 않을까~ 싶은 생각입니다!

hjsong333 replied:
공감합니다. 리팩토링 때 반영하겠습니다. 👍

bakhacode reviewed (commented):
전체적으로 클래스문법과 createElement를 사용으로 인한 특성에서 오는 불편함 외에는 따로 눈에 띄는 문제점 없이(다른 분들이 이미 캐치) 깔끔한 것 같아요!

hjsong333 committed:
refactor: dom creation and property in GridView

- Replacing raw DOM API with createElement in domUtil.
- Remove unused properties(tbody, rows, numbers).
- Removing unnecessary type of properties cause of type inference.
- Move addEventlistener code to setEvent method.

jsh3418 commented on src/components/NewsStand.ts:
```diff
@@ -0,0 +1,37 @@
+import Main from './main/Main';
+import NewsBar from './newsBar/NewsBar';
+
+type NewsStandProps = {
+  dateInfo: Date;
+  gridInfo: GridInfo;
+};
+
+type NewNewsStandProps = {
+  dateInfo: Date;
+  gridInfo: GridInfo;
+  subscriptionInfo: number[];
+};
```
interface 상속으로 중복을 줄여볼 수 있을 것 같아요!

jsh3418 commented on src/components/main/GridView.ts:
```diff
@@ -0,0 +1,207 @@
+    this.rightArrow = document.createElement('a');
+    this.rightArrow.href = '#';
+    this.rightArrow.classList.add(style.right_arrow);
+    if (this.page === 3) {
```
마지막 page가 3이 아닌 경우를 생각해줘야 할 것 같아요

hjsong333 committed:
refactor: 피드백 받은 사항 리팩토링

- 프로퍼티의 명시적 타입 선언을 타입 추론 방식으로 변경.
- as를 타입 가드로 대체.
- 커스텀 domUtil 함수로 Element 생성.

hjsong333 committed:
refactor: 인자로 받은 컴포넌트의 updateProps 호출

- 의미없는 props 드릴링을 없애기 위함.
- invoke 호출 시 재렌더링이 필요한 컴포넌트를 인자로 넘김.

hjsong333 committed:
refactor: 2가지 역할을 하는 함수를 2개의 함수로 분리

- getGridImgs에서 이미지 정보를 불러오고 섞는 2가지 역할을 가짐.
- shuffleArray라는 함수를 만들어 역할을 분리.
